Add LLVM pass timer

sklam opened the pull request:
Depends on https://github.com/numba/llvmlite/pull/624

- expose `llvm:: report_and_reset_timings()`
- provide python API to access the information

sklam commented:
CI is failing because we need a new llvmlite

stuartarchibald commented:
/AzurePipelines run

sklam commented:
/azp run

sklam commented:
Hmm... i'm surprised that the behavior of the timer is different depending on the OS.

stuartarchibald commented:
> Hmm... i'm surprised that the behavior of the timer is different depending on the OS.

Think there's some state going on somewhere, locally, on linux, this sometimes works and sometimes does not!

stuartarchibald commented:
Looks like sometimes the columns coming back from the timer are missing the "System Time" column.
```
===-------------------------------------------------------------------------===
                      ... Pass execution timing report ...
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0015 seconds (0.0015 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.0011 ( 80.3%)   0.0000 ( 30.4%)   0.0011 ( 77.9%)   0.0011 ( 77.8%)  SROA #3
   0.0001 (  9.5%)   0.0000 ( 34.8%)   0.0002 ( 10.7%)   0.0002 ( 10.6%)  Early CSE #2
   0.0001 (  5.6%)   0.0000 ( 21.7%)   0.0001 (  6.4%)   0.0001 (  6.2%)  Simplify the CFG #9
   0.0000 (  1.2%)   0.0000 (  2.9%)   0.0000 (  1.3%)   0.0000 (  1.3%)  Dominator Tree Construction #25
   0.0000 (  1.2%)   0.0000 (  4.3%)   0.0000 (  1.3%)   0.0000 (  1.3%)  Post-Dominator Tree Construction #5
   0.0000 (  0.9%)   0.0000 (  4.3%)   0.0000 (  1.0%)   0.0000 (  0.9%)  Dominator Tree Construction #26
   0.0000 (  0.4%)   0.0000 (  0.0%)   0.0000 (  0.3%)   0.0000 (  0.6%)  Instrument function entry/exit with calls to e.g. mcount() (pre inlining) #2
   0.0000 (  0.4%)   0.0000 (  1.4%)   0.0000 (  0.5%)   0.0000 (  0.5%)  Prune NRT refops #4
   0.0000 (  0.3%)   0.0000 (  0.0%)   0.0000 (  0.3%)   0.0000 (  0.4%)  Lower 'expect' Intrinsics #2
   0.0000 (  0.3%)   0.0000 (  0.0%)   0.0000 (  0.3%)   0.0000 (  0.3%)  Normalize NRT refops #4
   0.0014 (100.0%)   0.0001 (100.0%)   0.0015 (100.0%)   0.0015 (100.0%)  Total


[PassTimingRecord(user_time=0.0011, user_percent=80.3, system_time=0.0, system_percent=30.4, user_system_time=0.0011, user_system_percent=77.9, wall_time=0.0011, wall_percent=77.8, pass_name='SROA #3'), PassTimingRecord(user_time=0.0001, user_percent=9.5, system_time=0.0, system_percent=34.8, user_system_time=0.0002, user_system_percent=10.7, wall_time=0.0002, wall_percent=10.6, pass_name='Early CSE #2'), PassTimingRecord(user_time=0.0001, user_percent=5.6, system_time=0.0, system_percent=21.7, user_system_time=0.0001, user_system_percent=6.4, wall_time=0.0001, wall_percent=6.2, pass_name='Simplify the CFG #9'), PassTimingRecord(user_time=0.0, user_percent=1.2, system_time=0.0, system_percent=2.9, user_system_time=0.0, user_system_percent=1.3, wall_time=0.0, wall_percent=1.3, pass_name='Dominator Tree Construction #25'), PassTimingRecord(user_time=0.0, user_percent=1.2, system_time=0.0, system_percent=4.3, user_system_time=0.0, user_system_percent=1.3, wall_time=0.0, wall_percent=1.3, pass_name='Post-Dominator Tree Construction #5'), PassTimingRecord(user_time=0.0, user_percent=0.9, system_time=0.0, system_percent=4.3, user_system_time=0.0, user_system_percent=1.0, wall_time=0.0, wall_percent=0.9, pass_name='Dominator Tree Construction #26'), PassTimingRecord(user_time=0.0, user_percent=0.4, system_time=0.0, system_percent=0.0, user_system_time=0.0, user_system_percent=0.3, wall_time=0.0, wall_percent=0.6, pass_name='Instrument function entry/exit with calls to e.g. mcount() (pre inlining) #2'), PassTimingRecord(user_time=0.0, user_percent=0.4, system_time=0.0, system_percent=1.4, user_system_time=0.0, user_system_percent=0.5, wall_time=0.0, wall_percent=0.5, pass_name='Prune NRT refops #4'), PassTimingRecord(user_time=0.0, user_percent=0.3, system_time=0.0, system_percent=0.0, user_system_time=0.0, user_system_percent=0.3, wall_time=0.0, wall_percent=0.4, pass_name="Lower 'expect' Intrinsics #2"), PassTimingRecord(user_time=0.0, user_percent=0.3, system_time=0.0, system_percent=0.0, user_system_time=0.0, user_system_percent=0.3, wall_time=0.0, wall_percent=0.3, pass_name='Normalize NRT refops #4'), PassTimingRecord(user_time=0.0014, user_percent=100.0, system_time=0.0001, system_percent=100.0, user_system_time=0.0015, user_system_percent=100.0, wall_time=0.0015, wall_percent=100.0, pass_name='Total')]
===-------------------------------------------------------------------------===
                      ... Pass execution timing report ...
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0005 seconds (0.0005 wall clock)

   ---User Time---   --User+System--   ---Wall Time---  --- Name ---
   0.0001 ( 31.8%)   0.0001 ( 31.8%)   0.0001 ( 31.5%)  Simplify the CFG #10
   0.0001 ( 29.7%)   0.0001 ( 29.7%)   0.0001 ( 29.6%)  Early CSE #3
   0.0001 ( 17.1%)   0.0001 ( 17.1%)   0.0001 ( 16.9%)  SROA #4
   0.0000 (  5.6%)   0.0000 (  5.6%)   0.0000 (  5.6%)  Post-Dominator Tree Construction #6
   0.0000 (  5.1%)   0.0000 (  5.1%)   0.0000 (  5.3%)  Dominator Tree Construction #27
   0.0000 (  3.8%)   0.0000 (  3.8%)   0.0000 (  3.9%)  Dominator Tree Construction #28
   0.0000 (  2.4%)   0.0000 (  2.4%)   0.0000 (  2.4%)  Prune NRT refops #5
   0.0000 (  1.5%)   0.0000 (  1.5%)   0.0000 (  2.2%)  Instrument function entry/exit with calls to e.g. mcount() (pre inlining) #3
   0.0000 (  1.5%)   0.0000 (  1.5%)   0.0000 (  1.4%)  Lower 'expect' Intrinsics #3
   0.0000 (  1.5%)   0.0000 (  1.5%)   0.0000 (  1.3%)  Normalize NRT refops #5
   0.0005 (100.0%)   0.0005 (100.0%)   0.0005 (100.0%)  Total


[]
```
as a result the regex doesn't match and there's no records to parse.

stuartarchibald commented on docs/source/developer/llvm_timings.rst:
```diff
@@ -0,0 +1,105 @@
+Getting LLVM Pass Timings
+-------------------------
+
+The dispatcher stores LLVM pass timings in the metadata under the
```
```suggestion
The dispatcher stores LLVM pass timings in the dispatcher object metadata under the
```
and reference to this: https://numba.readthedocs.io/en/stable/reference/jit-compilation.html#Dispatcher.get_metadata

stuartarchibald commented on docs/source/developer/llvm_timings.rst:
```diff
@@ -0,0 +1,105 @@
+-------------------------
+
+The dispatcher stores LLVM pass timings in the metadata under the
+``llvm_pass_timings`` key. The timings info contains details on how much time
```
```suggestion
``llvm_pass_timings`` key. The timings information contains details on how much time
```

stuartarchibald replied:
Also wonder if this should be `llvm_pass_times` instead of `llvm_pass_timings` so as to be consistent with respect to the already present `pipeline_times` that describes similar.

Same applies throughout, should it be `s/Timings/Times` ?

stuartarchibald commented on docs/source/developer/llvm_timings.rst:
```diff
@@ -0,0 +1,105 @@
+
+The dispatcher stores LLVM pass timings in the metadata under the
+``llvm_pass_timings`` key. The timings info contains details on how much time
+has spent for each pass. The pass timings are also grouped by their purpose.
```
```suggestion
has spent been spent in each pass. The pass timings are also grouped by their purpose.
```

stuartarchibald commented on docs/source/developer/llvm_timings.rst:
```diff
@@ -0,0 +1,105 @@
+
+.. code-block:: text
+
+    Printing pass timings for JITCodeLibrary('DocsLLVMPassTimings.test_pass_timings.<locals>.foo')
```
It feels like this section would benefit from some explanation for what the output is showing. I think `#0, #1, #2` are for function passes running on the function, cpython wrapper and cfunc respectively (aside: why is cfunc on by default?). Then `#3` and `#4` are module level passes for the cheap and full runs, but I don't think this split compilation is explained anywhere in the docs? `#5` is finalization which makes sense!

sklam replied:
I don't think we need too much explanation here. It is reflecting what is happening and I expect things will change as we refine the optimizations later. I'd rather add more descriptive names for each pass group.

stuartarchibald commented on docs/source/developer/repomap.rst:
```diff
@@ -255,7 +255,8 @@ Misc Support
 - :ghfile:`numba/core/caching.py` - Disk cache for compiled functions
 - :ghfile:`numba/np/npdatetime.py` - Helper functions for implementing NumPy
   datetime64 support
-
```
Does a blank line need adding back in?

sklam replied:
It doesn't need two lines. I checked the rendered doc.

stuartarchibald commented on numba/core/codegen.py:
```diff
@@ -570,22 +577,29 @@ def _optimize_functions(self, ll_module):
-            self._final_module = remove_redundant_nrt_refct(self._final_module)
-        # The full optimisation suite is then run on the refop pruned IR
-        self._codegen._mpm_full.run(self._final_module)
+            with self._recorded_timings.record("Legacy refprune"):
+                self._final_module = remove_redundant_nrt_refct(
+                    self._final_module,
+                )
```
Not currently tested?

stuartarchibald commented on numba/core/codegen.py:
```diff
@@ -535,12 +536,18 @@ def __init__(self, codegen: "BaseCPUCodegen", name: str):
         # Track names of the dynamic globals
         self._dynamic_globals = []
+        ptc_name = f"{self.__class__.__name__}({self._name!r})"
+        self._recorded_timings = PassTimingsCollection(ptc_name)
```
Is there a performance impact for always collecting results?

sklam replied:
The maximum slowdown I measured is a 1.2x. The bigger the function, the less significant the slowdown is. I have a big function with a 1.06x slowdown. All measured with ipython timeit.

stuartarchibald commented on numba/misc/llvm_pass_timings.py:
```diff
@@ -0,0 +1,364 @@
+                      key=lambda x: x.timings.get_total_time(),
+                      reverse=True)
+
+    def summary(self):
```
Should this get a `n=` for how many records to dump?

stuartarchibald reviewed (commented):
Thanks for the patch, this looks good in general, just a few minor comments to resolve. Thanks again!

sklam commented:
I should have addressed all comments. On https://github.com/numba/numba/pull/6560#discussion_r552181599, does the slowdown worth adding a env-var to disable the timings?

stuartarchibald commented:
> I should have addressed all comments. On [#6560 (comment)](https://github.com/numba/numba/pull/6560#discussion_r552181599), does the slowdown worth adding a env-var to disable the timings?

Thanks @sklam. RE slowdown, I think this should just be off by default and enabled by env var or some other means. I don't think it's a cost that can be justified on the most common path, what do you think?

stuartarchibald commented on numba/misc/llvm_pass_timings.py:
```diff
@@ -327,8 +327,20 @@ def list_longest_first(self):
+
+        Parameters
+        ----------
+        topn : int; optional
```
```suggestion
        topn : int; optional, default=5.
```

sklam commented:
Added NUMBA_LLVM_PASS_TIMINGS to enable/disable the pass timings.

stuartarchibald commented on numba/tests/test_llvm_pass_timings.py:
```diff
@@ -1,11 +1,19 @@
 
 
 class TestLLVMPassTimings(TestCase):
+    def setUp(self):
+        self.__old_LLVM_PASS_TIMINGS = config.LLVM_PASS_TIMINGS
+        config.LLVM_PASS_TIMINGS = 1
+
+    def tearDown(self):
+        config.LLVM_PASS_TIMINGS = self.__old_LLVM_PASS_TIMINGS
+
```
Same comment as above RE `numba.tests.support.override_config`

stuartarchibald commented on numba/misc/llvm_pass_timings.py:
```diff
@@ -0,0 +1,400 @@
+
+        Parameters
+        ----------
+        topn : int; optional
```
This pattern renders strangely, in this case it appears as `topnint`: https://numba--6560.org.readthedocs.build/en/6560/developer/llvm_timings.html#api-for-custom-analysis

stuartarchibald reviewed (commented):
Thanks for adding the config variable. This looks good, few minor comments, I've fixed them in https://github.com/stuartarchibald/numba/commit/b0469dc6970568d167cfcf0b53710ed466866b2a if you are happy to accept a patch.

sklam commented:
@stuartarchibald I've accepted the patch.

stuartarchibald reviewed (approved):
Thanks for working on this, looks good.

stuartarchibald commented:
> @stuartarchibald I've accepted the patch.

Great, thanks, approved for merge.
